File: 366-find-leaves-of-binary-tree/366-find-leaves-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> findLeaves(TreeNode root) {
        List<List<Integer>> ans = new ArrayList();
        getHeight(root, ans);
        return ans;
    }
    
    private int getHeight(TreeNode root, List<List<Integer>> ans) {
        if (root == null) return -1;
        
        int left = getHeight(root.left, ans);
        int right = getHeight(root.right, ans);
        
        int height = Integer.max(left, right) + 1;
        if (height >= ans.size()) {
            ans.add(new ArrayList());
        }
        ans.get(height).add(root.val);
        return height;
        
    }
}

File: 366-find-leaves-of-binary-tree/366-find-leaves-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def findLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root: return []
        ans = []
        
        def findLevel(node):
            level = 0
            if node.left:
                level = max(level, 1+findLevel(node.left))
            if node.right:
                level = max(level, 1+findLevel(node.right))
                
            if level <len(ans):
                ans[level].append(node.val)
            else:
                ans.append([node.val])
            
            return level
        
        findLevel(root)
        return ans